Safer slot arithmetic.

jonathanknowles opened the pull request:
# Issue Number

#466 

# Overview

This change makes `slotAt` return `Nothing` if the given time isn't covered by a slot. Thus `slotAt` is prevented from creating an invalid `SlotId` that occurs before the start of the blockchain.
    
In addition:
    
* Function `slotFloor` now returns `Nothing` if the specified time occurs before the start of the blockchain.
* Function `slotCeiling` now returns the genesis slot if the specified time occurs before the start of the blockchain.
* We provide a new function `slotRange` that converts an inclusive time range into an inclusive slot range, returning `Nothing` if the range ends before the start of the blockchain.
* We use `slotRange` in the implementation of `listTransactions`.

jonathanknowles committed:
Make `slotAt` return `Nothing` if the given time isn't covered by a slot.

This change prevents `slotAt` from creating an invalid `SlotId` that
occurs before the start of the blockchain.

This change also has a few knock-on effects:

* Function `slotFloor` now returns `Nothing` if the specified time
  occurs before the start of the blockchain.

* Function `slotCeiling` now returns the genesis slot if the
  specified time occurs before the start of the blockchain.

We also:

* provide a new function `slotRange` that converts an inclusive time
  range into an inclusive slot range, returning `Nothing` if the range
  ends before the start of the blockchain.

* use `slotRange` in the implementation of `listTransactions`.

jonathanknowles committed:
Add function `slotMinBound`.

jonathanknowles committed:
Add `Arbitrary` instance for `SlotParameters`.

jonathanknowles committed:
Add basic property tests for `slotCeiling` and `slotFloor`.

jonathanknowles committed:
Add basic property test for `slotRange`.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -765,26 +764,30 @@ newWalletLayer tracer bp db nw tl = do
-        liftIO $ assemble tip
-            <$> DB.readTxHistory db (PrimaryKey wid) order range
+    _listTransactions wid mStart mEnd order =
+        maybe (pure []) listTransactionsWithinRange =<< getSlotRange
```
So if the range is too low, we return `[]` directly. I slightly fear this is scattering the filtering-logic.

A pure `Range UTCTime -> Range SlotId`, mapping would be simpler. But that would require mapping some `Range UTCTime` to an _empty_ `Range SlotId`, which cannot express.

I'm wondering what would happen if we replaced `Range a = Range (Maybe a) (Maybe a)` with say `Filter a = [SomeSymbolicConditionType]`. Particularly if we'd like different filtering capabilities in the future for less complicated pagination/batching. Actually… that would probably not work easily… 🤔 

Maybe never mind.

jonathanknowles replied:
@Anviking wrote:

> So if the range is too low, we return `[]` directly. I slightly fear this is scattering the filtering-logic.

I agree with you here.

The core problem is that we're unable to express the _empty_ range, since both lower and upper bounds are _inclusive_.  Therefore, we must resort to tricks like `Maybe (Range a)`.

However, if we change `Range` so that it _is_ possible to express the empty range, then we can just pass the empty range into the `listTransactionsWithRange` function.

I propose a simple change to accommodate this: make the upper bound of `Range` exclusive. This way, the range `Range (Just n) (Just n)` is _empty_ for all values of `n`.

Of course, we _could_ do something more general like allowing each bound to be _either_ inclusive _or_ exclusive, but I think we can get away without doing that for the moment. If we need to make it more general later on, we can always do that.

What do you think?

Anviking replied:
I think the reason we had inclusive bounds was so that clients always could fetch the next page without gaps, despite slotId/insertedAt not being unique, even if that meant that it would have to prune duplicates at the edges. (😐)

With `[a,b)` I think this would be fine for `Ascending` but not fine for `Descending`.

I think what you have in the PR is very fine, and we should stick with it unless we have an obvious alternative.

jonathanknowles replied:
> I think the reason we had inclusive bounds was so that clients always could fetch the next page without gaps, despite slotId/insertedAt not being unique, even if that meant that it would have to prune duplicates at the edges. (neutral_face)
> 
> With `[a,b)` I think this would be fine for `Ascending` but not fine for `Descending`.
> 
> I think what you have in the PR is very fine, and we should stick with it unless we have an obvious alternative.

Okay, in that case, I'll polish up this PR so that it's ready for final review.

We can think about a more general solution later on. :)

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -941,6 +991,22 @@ slotAt (SlotParameters (EpochLength el) (SlotLength sl) (StartTime st)) t =
+    slotStart =
+        slotCeiling sps <$> mStart
+    slotEnd =
+        maybe (Just Nothing) (fmap Just . slotFloor sps) mEnd
```
Is this three nested `Maybe`s, and could it be avoided?

jonathanknowles replied:
I'm not sure it's possible to avoid this. But feel free to suggest something better/clearer!

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -368,7 +370,7 @@ instance FromText SortOrder where
     { rStart :: Maybe a
     , rEnd :: Maybe a
-    } deriving (Eq, Show)
+    } deriving (Eq, Functor, Show)
```
Should probably replace this with a `mapRange` function, to avoid the situation where `rStart > rEnd`.

Anviking reviewed (commented):
Two passing thoughts, but looks great!

Also, maybe there is some missing, over-arching property, regarding `slotRange` and `slotStartTime` closer to business-requirements 🤔

jonathanknowles committed:
Use `inclusiveLowerBound` and `inclusiveUpperBound`.

These terms are clearer than `rStart` and `rEnd`, and make client code more
self-documenting.

jonathanknowles committed:
Improve comment for `Range` data type.

jonathanknowles committed:
Simplify boundary tests for `Range` type.

jonathanknowles committed:
Remove unnecessary comments.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -352,23 +354,32 @@ instance ToText SortOrder where
+-- | Value                           | Range       | Membership                |
+-- |                                 | Represented | Function                  |
+-- +=================================+=============+===========================+
+-- | @'Range' ('Just' x) ('Just' y)@ | @[ x, y ]@  | @\\p -> p >= x && p <= y@ |
```
wow, nice! A shame the syntax makes it a bit messier to read un-rendered though

jonathanknowles replied:
> wow, nice! A shame the syntax makes it a bit messier to read un-rendered though

I agree, it's a shame that we have to choose between one or the other.

For me, the most irritating bits are:

* The `@` symbol
* The need to escape the backslash "lambda" symbol.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -352,23 +354,32 @@ instance ToText SortOrder where
-    { rStart :: Maybe a
-    , rEnd :: Maybe a
-    } deriving (Eq, Show)
+    { inclusiveLowerBound :: Maybe a
```
👍

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -352,23 +354,32 @@ instance ToText SortOrder where
+-- +=================================+=============+===========================+
+-- | @'Range' ('Just' x) ('Just' y)@ | @[ x, y ]@  | @\\p -> p >= x && p <= y@ |
+-- +---------------------------------+-------------+---------------------------+
+-- | @'Range' ('Just' x) 'Nothing' @ | @[ x, ∞ ]@  | @\\p -> p >= x          @ |
```
```suggestion
-- | @'Range' ('Just' x) 'Nothing' @ | @[ x, ∞ )@  | @\\p -> p >= x          @ |
```

I thought using exclusive bounds with infinities were the common practice?

jonathanknowles replied:
Well spotted. I didn't realize that this was the case.
